### PR TITLE
Fix schema tracking for nested messages

### PR DESCRIPTION
**Pull request by piochelepiotr**
# What Does This Do

Adds recursive protection to only get the schema of the top level message.
This was not only causing CPU overhead, but was also producing wrong data (schema name was the name of the most nested message).

I believe we have the same issue with Avro. I will open later for it once I'm able to reproduce the bug in Avro.

# Motivation

# Additional Notes

# Contributor Checklist

- [ ] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [ ] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [ ] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [ ] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [ ] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [PROJ-IDENT]

### CODE REVIEW

File: dd-java-agent/instrumentation/protobuf/src/test/groovy/readme.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+To generate the Protobuf message, we need to use an old version of protobuf (3.0.0), because we want to support the protobuf library starting at 3.0.0
```
Comment (piochelepiotr) [inline dd-java-agent/instrumentation/protobuf/src/test/groovy/readme.md:1]: I struggled with that for a while ... again (I already struggled with it when I added this integration).
So I decided to add a small readme, even if it's just for future me.

Comment (PerfectSlayer) [reply dd-java-agent/instrumentation/protobuf/src/test/groovy/readme.md:1]: Can’t we make it part of the build script by creating a custom gradle task for?
I did it in my previous company and worked well. You can even pick the version of protoc you would like to use.

**Review by PerfectSlayer — COMMENTED** [review]
I was more thinking of reusing [the Gradle plugin issued by Google](https://github.com/google/protobuf-gradle-plugin). That would prevent to re-invent the wheel.

In addition, everything could stay within the instrumentation gradle module, and the generation task should directly be part of the build `task`.

**Review by PerfectSlayer — APPROVED** [review]
Looks good from module change / packaging and test.
You might want another pair of eyes from @DataDog/apm-idm-java for the `CallDepthThreadLocalMap` usage 👀

**Review by nayeem-kamal — APPROVED** [review]

**Pull request merged**